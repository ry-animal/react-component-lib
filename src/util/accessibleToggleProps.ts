export function accessibleToggleProps(
  label: string,
  controls: string,
  expanded: boolean,
  popupType?: "menu"
) {
  return {
    "aria-controls": controls,
    "aria-expanded": expanded,
    "aria-haspopup": popupType || true,
    "aria-label": label,
    "aria-pressed": expanded,
    role: "button",
  };
}
