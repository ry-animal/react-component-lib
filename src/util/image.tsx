import { ISelectedFile } from "src/components/Forms/FileUpload/ISelectedFile";

export const makeDataUrl = (selectedFile: ISelectedFile) =>
  URL.createObjectURL(selectedFile.file);
