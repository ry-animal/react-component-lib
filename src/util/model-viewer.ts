/*
 * This file is a hacky way of allowing any other component to
 * use the <model-viewer /> tag just by doing:
 *   import "../../../util/model-viewer";
 *
 * The authors of the package do not expose it in a React-friendly
 * way, see: https://github.com/google/model-viewer/issues/1502
 *
 * The component documentation is at https://modelviewer.dev/docs/
 */

import "@google/model-viewer";

declare global {
  // eslint-disable-next-line @typescript-eslint/no-namespace
  namespace JSX {
    export interface IntrinsicElements {
      "model-viewer": {
        src?: string;
        alt?: string;
        poster?: string;
        autoplay?: boolean;
        exposure?: number;
        "camera-controls"?: boolean;
        "environment-image"?: string;
        "seamless-poster"?: string;
        loading?: "auto" | "lazy" | "eager";
        reveal?: "auto" | "interaction" | "manual";
      } & React.DetailedHTMLProps<
        React.HTMLAttributes<HTMLElement>,
        HTMLElement
      >;
    }
  }
}
