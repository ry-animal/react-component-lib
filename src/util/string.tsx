const DEFAULT_LOCALE = "en-US";

export const truncateEthAddress = (str: string): string => {
  if (str.length) {
    const numStartingCharacters = 6;
    const numTrailingCharacters = 4;
    return `${str.substring(0, numStartingCharacters)}...${str.substring(
      str.length - numTrailingCharacters,
      str.length
    )}`;
  }
  return "";
};

export const formatUSDPriceString = (value: number): string => {
  const minimumFractionDigits = 2;
  const maximumFractionDigits = 2;
  return `$${value.toLocaleString(DEFAULT_LOCALE, {
    minimumFractionDigits,
    maximumFractionDigits,
  })}`;
};

export const formatETHPriceString = (priceInEth: number): string => {
  const minimumFractionDigits = 0;
  const maximumFractionDigits = 8;
  const maxDigitsWithDecimal = 8;

  let formatted = `${priceInEth
    .toLocaleString(DEFAULT_LOCALE, {
      minimumFractionDigits,
      maximumFractionDigits,
    })
    .replace(/,/gu, "")
    .substring(0, maxDigitsWithDecimal)}`;

  if (formatted.endsWith(".")) {
    formatted = formatted.slice(0, -1);
  }
  return formatted;
};

const evaluateNumberLessThanOne = (num: number) => {
  if (num === 0) {
    return "0";
  }

  const minDecimalValue = 0.0001;
  if (num < minDecimalValue) {
    return minDecimalValue.toString();
  }
  // To round decimal to 4 digits
  const maxDigits = 10000;
  return `${Math.round(num * maxDigits) / maxDigits}`;
};

// Abbreviate Number https://stackoverflow.com/a/9462382

export const abbreviateNumber = (num: number): string => {
  const lookup = [
    { value: 1, symbol: "" },
    { value: 1e3, symbol: "k" },
    { value: 1e6, symbol: "M" },
    { value: 1e9, symbol: "B" },
    { value: 1e12, symbol: "T" },
    { value: 1e15, symbol: "P" },
    { value: 1e18, symbol: "E" },
  ];

  const rx = /\.0+$|(?<id>\.[0-9]*[1-9])0+$/u;
  const digits = 1;

  const item = lookup
    .slice()
    .reverse()
    .find((x) => num >= x.value);
  return item
    ? (num / item.value).toFixed(digits).replace(rx, "$1") + item.symbol
    : evaluateNumberLessThanOne(num);
};

export const getDomainFromRegex = (url: string) => {
  const matches = url.match(/:\/\/(?<id>.[^/]+)/u);
  // Extract domain (will be null if no match is found)
  return matches && matches[1];
};
