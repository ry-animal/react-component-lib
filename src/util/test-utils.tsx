import { render, RenderOptions } from "@testing-library/react";
import "jest-styled-components";
import * as React from "react";
import { BrowserRouter } from "react-router-dom";
import { ThemeProvider } from "styled-components";
import { LIGHT_THEME } from "../theme";

interface ProvidersProps {
  children: React.ReactNode;
}

function Providers({ children }: ProvidersProps) {
  return (
    <BrowserRouter>
      <ThemeProvider theme={LIGHT_THEME}>{children}</ThemeProvider>
    </BrowserRouter>
  );
}

const customRender = (
  ui: JSX.Element,
  options?:
    | RenderOptions<
        typeof import("@testing-library/dom/types/queries"),
        HTMLElement
      >
    | undefined
) => render(ui, { wrapper: Providers, ...options });

// Duplicated instead of import "../../__mocks__/matchMediaMock"; because jest has trouble resolving imports
Object.defineProperty(window, "matchMedia", {
  writable: true,
  value: jest.fn().mockImplementation((query) => ({
    matches: false,
    media: query,
    onchange: null,
    addListener: jest.fn(),
    removeListener: jest.fn(),
    addEventListener: jest.fn(),
    removeEventListener: jest.fn(),
    dispatchEvent: jest.fn(),
  })),
});

// Re-export everything
export * from "@testing-library/react";
// Override render method
export { customRender as render };
