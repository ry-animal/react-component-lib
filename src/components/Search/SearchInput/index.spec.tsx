import userEvent from "@testing-library/user-event";
import { act } from "react-dom/test-utils";
import { SearchInput } from ".";
import { render, screen } from "../../../util/test-utils";

describe("<SearchInput />", () => {
  describe("rendering", () => {
    test("should render according to snapshot (default)", () => {
      expect(render(<SearchInput />).asFragment()).toMatchSnapshot();
    });

    test("should render according to snapshot (backgroundColor)", () => {
      expect(
        render(<SearchInput backgroundColor="yellow" />).asFragment()
      ).toMatchSnapshot();
      expect(screen.getByRole("search")).toHaveStyleRule(
        "background-color",
        "yellow"
      );
    });

    test("should render according to snapshot (width)", () => {
      expect(
        render(<SearchInput width="10px" />).asFragment()
      ).toMatchSnapshot();
      expect(screen.getByRole("search")).toHaveStyleRule("width", "10px");
    });
  });

  describe("event handling", () => {
    test("onEscape called when esc pressed", async () => {
      const onEscape = jest.fn();
      render(<SearchInput onEscape={onEscape} />);
      await act(async () => {
        await screen.findByRole("search");
        userEvent.keyboard("{Esc}");
      });
      expect(onEscape).toHaveBeenCalledTimes(1);
    });
  });
});
