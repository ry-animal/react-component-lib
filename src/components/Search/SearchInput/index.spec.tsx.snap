// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SearchInput /> rendering should render according to snapshot (backgroundColor) 1`] = `
<DocumentFragment>
  .c1 {
  background-color: yellow;
  color: #151515;
  border-radius: 4px;
  font-size: 14px;
  padding-left: 48px;
  padding-right: 12px;
  width: 100%;
  -webkit-backdrop-filter: blur(8px);
  backdrop-filter: blur(8px);
  background-position: 15px center;
  background-repeat: no-repeat;
  border-width: 0;
  height: 48px;
}

.c1::-webkit-input-placeholder {
  color: #A4A4A4;
}

.c1::-moz-placeholder {
  color: #A4A4A4;
}

.c1:-ms-input-placeholder {
  color: #A4A4A4;
}

.c1::placeholder {
  color: #A4A4A4;
}

.c1:focus {
  outline: 2px solid #404040;
}

.c2 {
  position: absolute;
  top: 16px;
  left: 16px;
  height: 16px;
  width: 16px;
}

.c0 {
  position: relative;
}

<div
    class="c0"
  >
    <input
      class="c1"
      placeholder="Search GameStop NFT"
      role="search"
      type="search"
      width="100%"
    />
    <div
      class="c2"
    >
      <svg
        fill="none"
        height="17"
        viewBox="0 0 17 17"
        width="17"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          clip-rule="evenodd"
          d="M10.9212 12.0255C9.7923 12.9031 8.37376 13.4256 6.83317 13.4256C3.15127 13.4256 0.166504 10.4409 0.166504 6.75895C0.166504 3.07705 3.15127 0.0922852 6.83317 0.0922852C10.5151 0.0922852 13.4998 3.07705 13.4998 6.75895C13.4998 8.29954 12.9773 9.71809 12.0997 10.847L16.5891 15.3364C16.9145 15.6618 16.9145 16.1894 16.5891 16.5149C16.2637 16.8403 15.736 16.8403 15.4106 16.5149L10.9212 12.0255ZM11.8332 6.75895C11.8332 9.52038 9.59459 11.759 6.83317 11.759C4.07175 11.759 1.83317 9.52038 1.83317 6.75895C1.83317 3.99753 4.07175 1.75895 6.83317 1.75895C9.59459 1.75895 11.8332 3.99753 11.8332 6.75895Z"
          fill="#151515"
          fill-rule="evenodd"
        />
      </svg>
    </div>
  </div>
</DocumentFragment>
`;

exports[`<SearchInput /> rendering should render according to snapshot (default) 1`] = `
<DocumentFragment>
  .c1 {
  background-color: rgba(255,255,255,0.9);
  color: #151515;
  border-radius: 4px;
  font-size: 14px;
  padding-left: 48px;
  padding-right: 12px;
  width: 100%;
  -webkit-backdrop-filter: blur(8px);
  backdrop-filter: blur(8px);
  background-position: 15px center;
  background-repeat: no-repeat;
  border-width: 0;
  height: 48px;
}

.c1::-webkit-input-placeholder {
  color: #A4A4A4;
}

.c1::-moz-placeholder {
  color: #A4A4A4;
}

.c1:-ms-input-placeholder {
  color: #A4A4A4;
}

.c1::placeholder {
  color: #A4A4A4;
}

.c1:focus {
  outline: 2px solid #404040;
}

.c2 {
  position: absolute;
  top: 16px;
  left: 16px;
  height: 16px;
  width: 16px;
}

.c0 {
  position: relative;
}

<div
    class="c0"
  >
    <input
      class="c1"
      placeholder="Search GameStop NFT"
      role="search"
      type="search"
      width="100%"
    />
    <div
      class="c2"
    >
      <svg
        fill="none"
        height="17"
        viewBox="0 0 17 17"
        width="17"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          clip-rule="evenodd"
          d="M10.9212 12.0255C9.7923 12.9031 8.37376 13.4256 6.83317 13.4256C3.15127 13.4256 0.166504 10.4409 0.166504 6.75895C0.166504 3.07705 3.15127 0.0922852 6.83317 0.0922852C10.5151 0.0922852 13.4998 3.07705 13.4998 6.75895C13.4998 8.29954 12.9773 9.71809 12.0997 10.847L16.5891 15.3364C16.9145 15.6618 16.9145 16.1894 16.5891 16.5149C16.2637 16.8403 15.736 16.8403 15.4106 16.5149L10.9212 12.0255ZM11.8332 6.75895C11.8332 9.52038 9.59459 11.759 6.83317 11.759C4.07175 11.759 1.83317 9.52038 1.83317 6.75895C1.83317 3.99753 4.07175 1.75895 6.83317 1.75895C9.59459 1.75895 11.8332 3.99753 11.8332 6.75895Z"
          fill="#151515"
          fill-rule="evenodd"
        />
      </svg>
    </div>
  </div>
</DocumentFragment>
`;

exports[`<SearchInput /> rendering should render according to snapshot (width) 1`] = `
<DocumentFragment>
  .c1 {
  background-color: rgba(255,255,255,0.9);
  color: #151515;
  border-radius: 4px;
  font-size: 14px;
  padding-left: 48px;
  padding-right: 12px;
  width: 10px;
  -webkit-backdrop-filter: blur(8px);
  backdrop-filter: blur(8px);
  background-position: 15px center;
  background-repeat: no-repeat;
  border-width: 0;
  height: 48px;
}

.c1::-webkit-input-placeholder {
  color: #A4A4A4;
}

.c1::-moz-placeholder {
  color: #A4A4A4;
}

.c1:-ms-input-placeholder {
  color: #A4A4A4;
}

.c1::placeholder {
  color: #A4A4A4;
}

.c1:focus {
  outline: 2px solid #404040;
}

.c2 {
  position: absolute;
  top: 16px;
  left: 16px;
  height: 16px;
  width: 16px;
}

.c0 {
  position: relative;
}

<div
    class="c0"
  >
    <input
      class="c1"
      placeholder="Search GameStop NFT"
      role="search"
      type="search"
      width="10px"
    />
    <div
      class="c2"
    >
      <svg
        fill="none"
        height="17"
        viewBox="0 0 17 17"
        width="17"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          clip-rule="evenodd"
          d="M10.9212 12.0255C9.7923 12.9031 8.37376 13.4256 6.83317 13.4256C3.15127 13.4256 0.166504 10.4409 0.166504 6.75895C0.166504 3.07705 3.15127 0.0922852 6.83317 0.0922852C10.5151 0.0922852 13.4998 3.07705 13.4998 6.75895C13.4998 8.29954 12.9773 9.71809 12.0997 10.847L16.5891 15.3364C16.9145 15.6618 16.9145 16.1894 16.5891 16.5149C16.2637 16.8403 15.736 16.8403 15.4106 16.5149L10.9212 12.0255ZM11.8332 6.75895C11.8332 9.52038 9.59459 11.759 6.83317 11.759C4.07175 11.759 1.83317 9.52038 1.83317 6.75895C1.83317 3.99753 4.07175 1.75895 6.83317 1.75895C9.59459 1.75895 11.8332 3.99753 11.8332 6.75895Z"
          fill="#151515"
          fill-rule="evenodd"
        />
      </svg>
    </div>
  </div>
</DocumentFragment>
`;
